qiskit_experiments/framework/cache_method.py

```diff
@@ -0,0 +1,144 @@
+
+    Returns:
+        The decorator for caching methods.
+    """
+    cache_fn = _cache_function(cache)
+    cache_key_fn = _cache_key_function(cache_args, require_hashable)
+
+    def cache_method_decorator(method: Callable) -> Callable:
+        """Decorator for caching method.
+
+        Args:
+            method: A method to cache.
+
+        Returns:
+            The wrapped cached method.
+        """
+
+        @functools.wraps(method)
+        def _cached_method(self, *args, **kwargs):
+            meth_cache = cache_fn(self, method)
+            key = cache_key_fn(*args, **kwargs)
+            if key in meth_cache:
+                return meth_cache[key]
+            result = method(self, *args, **kwargs)
+            meth_cache[key] = result
+            return result
+
+        return _cached_method
+
+    return cache_method_decorator
+
+
+def _cache_key_function(cache_args: bool, require_hashable: bool) -> Callable:
+    """Return function for generating cache keys.
+
+    Args:
+        cache_args: If True include method arg and kwarg values when
+                    caching the method. If False only a single return will
+                    be cached for the method regardless of any args.
+        require_hashable: If True require all cached args and kwargs are
+                          hashable. If False un-hashable values are allowed
+                          but will be excluded from the cache key.
+
+    Returns:
+        The functions for generating cache keys.
+    """
+    if not cache_args:
+
+        def _cache_key(*args, **kwargs):
+            # pylint: disable = unused-argument
+            return tuple()
+
+    elif require_hashable:
+
+        def _cache_key(*args, **kwargs):
+            return args + tuple(list(kwargs.items()))
+
+    else:
+
+        def _cache_key(*args, **kwargs):
+            cache_key = tuple()
+            for arg in args:
+                try:
+                    hash(arg)
+                except TypeError:
+                    continue
+                cache_key += (arg,)
+            for key, value in kwargs.items():
+                try:
+                    hash(value)
+                except TypeError:
+                    continue
+                cache_key += ((key, value),)
+            return cache_key
+
+    return _cache_key
+
+
+def _cache_function(cache: Union[Dict, str]) -> Callable:
+    """Return function for initializing and accessing cache dict.
+
+    Args:
+        cache: The cache or cache attribute name to use. If a dict it will
+               be used directly, if a str a cache dict will be created under
+               that attribute name if one is not already present.
+
+    Returns:
+        The function for accessing the cache dict.
+    """
+    if isinstance(cache, str):
+
+        def _cache_fn(instance, method):
+            if not hasattr(instance, cache):
+                setattr(instance, cache, {})
```

Comment:
You can share caches across instances if you want by using it with an existing dict instead of a string (see `test_cache_dict`). 

LRU cache is defined for a single function. If you want to share a cached function (called by a method) across classes you can continue to use lru_cache.

Reply:
Make sense. Now I understand the use case of the dictionary.